languages/python/reference/exercise-concepts/robot-simulator.md

```diff
@@ -0,0 +1,62 @@
+# Concepts for robot-simulator
+
+[Example Implementation](https://github.com/exercism/python/blob/master/exercises/robot-simulator/example.py)
+
+
+## General
+
+- contstants: Are not enforced by the runtime, but are used via the convention of `<ALL_CAPS_NAMING>` to signal that these values are expected to remaine unchanged.  Perferably, constants are defined at a module level. The example solution uses the ALL_CAPS convention to define NORTH, SOUTH, EAST, and WEST constants. 
+
+- multiple assignment:  Python allows multiple assignment, assiging the items on the left of `=` _in order_ to the values on the right of `=`.  Equivelent to forming tuples of each side and then "unpacking" them.  This exercise solution uses multiple assignment for the contant values NORTH, EAST, SOUTH, WEST.
+
+- `range()` built-in type:  Represents an immutable sequence of numbers (or any object that implements the `__index__` magic method).  Used in the example to represnet the values from zero to 3 as assigned to NORTH, EAST, SOUTH, WEST. 
+
+- classes: The exercise objective is to define a `robot` type.  Tested methods are linked to a `robot` class.
+
+- objects: Creating different instances of the `robot` class with different data representing different starting positions and bearing are tested.
+
+- constructor: Customizing object initalization with actions and persisting data. The example uses a constructor (via `__init__` + `__new__`) to persist a `compass` object and x, y coordinates assigned to instance properties.
```

Comment:
The issue with combining `__new__` and __init__` are that the first one is the actual class _constructor_, the second is the _initializer_ ... the distinction is often meaningless in Python because we so rarely subclass builtin immutable types, but nevertheless there's still a difference.

Reply:
I struggle with this, because the closest concept is _constructor_ for other OOP languages but `__init__` _initializes_.  Should we be calling out `__new__` when it's most often called under the covers?   I changed this to "initialization", but I feel like that might confuse learners - but "constructor" doesn't sound fully accurate either.

> - initialization: Customizing object instatiation with actions and persisting data. The example uses `__init__` to persist a `compass` object and x, y coordinates assigned to instance attributes.

Reply:
New text spelling "instatiation" -> "instantiation". Might be better as "The example uses `__init__` to take a newly constructed `compass` instance and assign **x** and **y** coordinates to it as persistent attributes" or something similar.

Reply:
And yeah it's a bit of a weird, multi-step concept in Python ...  for most people's purposes the class constructor is, semantically, `c = Class()`, but that _cannot_ be just `__init__`, right, since `__init__` never returns _anything_ to assign to **c**? I don't know how far we need to dive into the data model (do we need or want to explain that `Class()` is actually `type.__call__(Class)` via the metaclass assigned by default to `Class`, and that it's inside the implementation of _that_ we find the algorithm that uses `__new__` to create **self**, `__init__` to mutate it, and _then_ returns **self**?) or do we just save explaining `__new__` for an exercise that subclasses an immutable builtin?